File: AMF5/src/test/java/javaPlatform/TransformationTest.java
package javaPlatform;

import amf.apicontract.client.platform.AMFClient;
import amf.apicontract.client.platform.OASConfiguration;
import amf.apicontract.client.platform.RAMLConfiguration;
import amf.apicontract.client.platform.model.domain.EndPoint;
import amf.apicontract.client.platform.model.domain.Operation;
import amf.apicontract.client.platform.model.domain.api.WebApi;
import amf.core.client.platform.model.document.BaseUnit;
import amf.core.client.platform.model.document.Document;
import org.junit.Test;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.stream.Collectors;

import static org.junit.Assert.*;

public class TransformationTest {

    @Test
    public void resolveRaml10() throws ExecutionException, InterruptedException {
        final AMFClient client = RAMLConfiguration.RAML10().createClient();

        final BaseUnit unresolvedModel = client.parse("file://src/test/resources/examples/raml-resource-type.raml").get().baseUnit();
        assertNotNull(unresolvedModel);

        final BaseUnit resolvedModel = client.transform(unresolvedModel).baseUnit();
        assertNotNull(resolvedModel);
        final WebApi api = (WebApi) ((Document) resolvedModel).encodes();
        assertFalse(api.endPoints().get(0).operations().isEmpty());
    }

    @Test
    public void resolveOas30() throws ExecutionException, InterruptedException {
        final AMFClient client = OASConfiguration.OAS30().createClient();

        final BaseUnit unresolvedModel = client.parse("file://src/test/resources/examples/banking-api-oas30.json").get().baseUnit();
        assertNotNull(unresolvedModel);

        final BaseUnit resolvedModel = client.transform(unresolvedModel).baseUnit();
        assertNotNull(resolvedModel);

        final Document document = (Document) resolvedModel;
        final WebApi api = (WebApi) document.encodes();
        final List<Operation> operations = api.endPoints().stream().map(EndPoint::operations).flatMap(Collection::stream).collect(Collectors.toList());
        final List<Operation> operationWithoutServers = operations.stream().filter(op -> op.servers().isEmpty()).collect(Collectors.toList());
        assertTrue(operationWithoutServers.isEmpty());
        assertFalse(operations.isEmpty());
    }

    @Test
    public void resolveRamlOverlay() throws ExecutionException, InterruptedException {
        final AMFClient client = RAMLConfiguration.RAML10().createClient();

        final Document unresolvedModel = client.parseDocument("file://src/test/resources/examples/raml-overlay/test-overlay.raml").get().document();
        assertEquals("unresolved overlay should reference main API", 1, unresolvedModel.references().size());

        final Document resolvedModel = (Document) client.transform(unresolvedModel).baseUnit();
        assertEquals("resolved model shouldn't reference anything", 0, resolvedModel.references().size());
        final WebApi api = (WebApi) ((Document) resolvedModel).encodes();
        assertTrue(api.endPoints().size() > 1);

    }
}

File: AMF5/src/test/scala/scalaPlatform/TransformationTest.scala
package scalaPlatform

import amf.apicontract.client.scala.{OASConfiguration, RAMLConfiguration}
import amf.apicontract.client.scala.model.domain.api.WebApi
import amf.core.client.scala.model.document.Document
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.should

class TransformationTest extends AsyncFlatSpec with should.Matchers {

  "AMF transformation" should "transform a RAML 1.0 applying resource types with default pipeline" in {
    val client = RAMLConfiguration.RAML10().baseUnitClient()
    client.parse("file://src/test/resources/examples/raml-resource-type.raml") map { parseResult =>
      val transformed = client.transform(parseResult.baseUnit)
      val document = transformed.baseUnit.asInstanceOf[Document]
      val allOperations = document.encodes.asInstanceOf[WebApi].endPoints.flatMap(_.operations)
      assert(allOperations.nonEmpty,"resource type should be resolved defining new operation")
    }
  }

  it should "transform an OAS 3.0 api with default pipeline" in {
    val client = OASConfiguration.OAS30().baseUnitClient()
    client.parse("file://src/test/resources/examples/banking-api-oas30.json") map { parseResult =>
      val transformed =
        client.transform(parseResult.baseUnit)
      val document = transformed.baseUnit.asInstanceOf[Document]
      val allOperations = document.encodes.asInstanceOf[WebApi].endPoints.flatMap(_.operations)
      assert(allOperations.forall(_.servers.nonEmpty),"servers should be resolved to operations")
    }
  }

  it should "apply a RAML 1.0 Overlay to an api" in {
    val client = RAMLConfiguration.RAML10().baseUnitClient()
    client.parse(
      "file://src/test/resources/examples/raml-overlay/test-overlay.raml"
    ) map { parseResult =>
      assert(
        parseResult.baseUnit.references.size == 1,
        "unresolved overlay should reference main API"
      )
      val transformed =
        client.transform(parseResult.baseUnit)

      assert(
        transformed.baseUnit.references.isEmpty,
        "transformed model shouldn't reference anything"
      )
      val webapi = transformed.baseUnit.asInstanceOf[Document].encodes.asInstanceOf[WebApi]
      assert(webapi.endPoints.size > 1)
    }
  }
}
